rust/src/storage/file.rs

```diff
@@ -58,26 +63,14 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
-
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
-
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
-            .create_new(true)
-            .write(true)
-            .open(path)
-            .await
-            .map_err(|e| match e.kind() {
-                std::io::ErrorKind::AlreadyExists => StorageError::AlreadyExists(path.to_string()),
-                _ => StorageError::Io { source: e },
-            })?;
+        let tmp_file_name = format!("{}.temporary", Uuid::new_v4().to_string());
+        let tmp_path = self.join_path(&self.root, &tmp_file_name);
 
+        let mut f = fs::File::create(&tmp_path).await?;
         f.write(obj_bytes).await?;
 
+        super::rename::rename(&tmp_path, path)?;
```

Comment:
Since this rename is specific to fs backend, it would be better to namespace it within the file module so it's self contained. To do that, we can change `file.rs` into the following structure:

```
src/storage/file/mod.rs
src/storage/file/rename.rs
```

then from within `file/mod.rs`, we can import the rename code using `crate::storage::file::rename`.